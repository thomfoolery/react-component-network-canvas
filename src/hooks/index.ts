export * from "./useBridge";
export * from "./useDragManager";
export * from "./useGraphManager";
export * from "./useOptions";
export * from "./usePanZoom";
export * from "./useWorkspace";
