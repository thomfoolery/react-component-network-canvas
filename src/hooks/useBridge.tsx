import React, {createContext, useMemo, useContext, ReactNode} from "react";
import * as Types from "@component/types";

const Context = createContext();

interface Props {
  value?: Types.Bridge;
  children?: ReactNode;
}

const defaultBridge: Types.Bridge = {
  connect() {},
  onChangeZoom() {},
  onUpdateGraph() {},
  onClickCanvas() {},
  onClickPort() {},
  onKeyPress() {},
};

export function BridgeProvider(props: Props) {
  const {value, children} = props;
  const bridge = useMemo(() => {
    return {
      ...defaultBridge,
      ...value,
    };
  }, [value]);

  return <Context.Provider value={bridge}>{children}</Context.Provider>;
}

export function useBridge() {
  return useContext(Context);
}
