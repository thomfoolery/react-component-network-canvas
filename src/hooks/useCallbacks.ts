import { useContext } from "react";

import { CallbacksContext } from "@component/contexts";
import * as Types from "@component/types";

function useCallbacks(): Types.Callbacks {
  return useContext(CallbacksContext);
}

export { useCallbacks };
