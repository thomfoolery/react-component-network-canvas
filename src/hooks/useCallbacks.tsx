import React, { createContext, useMemo, useContext, ReactNode } from "react";
import * as Types from "@component/types";

const Context = createContext();

interface Props {
  value?: Types.Callbacks;
  children?: ReactNode;
}

const defaultCallbacks: Types.Callbacks = {
  onMount() {
    return undefined;
  },
  onChangeZoom() {
    return undefined;
  },
  onMutateGraph() {
    return undefined;
  },
  onClickCanvas() {
    return undefined;
  },
  onClickNode() {
    return undefined;
  },
  onClickPort() {
    return undefined;
  },
  onDropCanvas() {
    return undefined;
  },
  onKeyPress() {
    return undefined;
  },
  onChangeSelectedNodeIds() {
    return undefined;
  },
};

function createCallbacks(
  value: Partial<Types.Callbacks> = {}
): Types.Callbacks {
  return {
    ...defaultCallbacks,
    ...value,
  };
}

function CallbacksProvider(props: Props): ReactNode {
  const { value, children } = props;
  const bridge = useMemo(() => createCallbacks(value), [value]);

  return <Context.Provider value={bridge}>{children}</Context.Provider>;
}

function useCallbacks(): Types.Callbacks {
  return useContext(Context);
}

export { createCallbacks, useCallbacks, CallbacksProvider };
