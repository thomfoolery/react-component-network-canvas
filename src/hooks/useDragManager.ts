import { useContext } from "react";

import { DragManagerContext } from "@component/contexts";
import * as Types from "@component/types";

function useDragManager(): Types.DragManager {
  return useContext(DragManagerContext);
}

export { useDragManager };
