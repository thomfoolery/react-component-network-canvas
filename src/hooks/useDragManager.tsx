import React, {
  createContext,
  useRef,
  useMemo,
  useContext,
  ReactNode,
} from "react";
import {createPublisher} from "@component/utils";
import * as Types from "@component/types";

const Context = createContext();

interface DragManagerPrivateProps {
  dragData: any;
  dragDelta: Types.Position;
  dragStartPosition: Types.Position | null;
  workspace?: Types.Workspace;
  subscriptions: {
    dragStartById: Publisher;
    dragMoveById: Publisher;
    dragEndById: Publisher;
  };
}

function createDragManager() {
  const __: DragManagerPrivateProps = {
    dragData: null,
    dragDelta: {x: 0, y: 0},
    dragStartPosition: null,
    workspace: undefined,
    subscriptions: {
      dragStartById: createPublisher(),
      dragMoveById: createPublisher(),
      dragEndById: createPublisher(),
    },
  };

  function handleDragStart(event) {
    document.addEventListener("selectstart", handleOnSelectStart);
    event.currentTarget.addEventListener("mousemove", handleDragMove);
    __.dragStartPosition = {x: event.screenX, y: event.screenY};
    __.dragDelta = {x: 0, y: 0};
    __.subscriptions.dragStartById.notifyAll(event);
  }

  function handleDragMove(event) {
    if (!__.dragStartPosition) return;

    const {workspace, dragStartPosition} = __;
    const currentPosition = {x: event.screenX, y: event.screenY};
    const dragDelta = {
      x:
        (currentPosition.x - dragStartPosition.x) /
        (workspace?.panZoom.zoom || 1),
      y:
        (currentPosition.y - dragStartPosition.y) /
        (workspace?.panZoom.zoom || 1),
    };
    __.dragDelta = dragDelta;
    __.subscriptions.dragMoveById.notifyAll(event, dragDelta, __.dragData);
  }

  function handleDragEnd(event) {
    document.removeEventListener("selectstart", handleOnSelectStart);
    event.currentTarget.removeEventListener("mousemove", handleDragMove);
    if (__.dragStartPosition) {
      const {workspace, dragStartPosition} = __;
      const currentPosition = {x: event.screenX, y: event.screenY};
      const dragDelta = {
        x:
          (currentPosition.x - dragStartPosition.x) /
          (workspace?.panZoom.zoom || 1),
        y:
          (currentPosition.y - dragStartPosition.y) /
          (workspace?.panZoom.zoom || 1),
      };
      __.subscriptions.dragEndById.notifyAll(event, dragDelta, __.dragData);
    }
    __.dragStartPosition = null;
    __.dragData = null;
  }

  return {
    set workspace(workspace: Types.Workspace) {
      __.workspace = workspace;
    },
    get dragDelta() {
      return {...__.dragDelta};
    },
    get dragData() {
      return {...__.dragData};
    },
    set dragData(dragData: any) {
      __.dragData = dragData;
    },

    // event handlers
    handleDragStart,
    handleDragMove,
    handleDragEnd,

    // subscriptions
    subscribeToDragStart(id, fn) {
      __.subscriptions.dragStartById.addListenerForId(id, fn);
    },
    unsubscribeToDragStart(id, fn) {
      __.subscriptions.dragStartById.removeListenerForId(id, fn);
    },
    subscribeToDragMove(id, fn) {
      __.subscriptions.dragMoveById.addListenerForId(id, fn);
    },
    unsubscribeToDragMove(id, fn) {
      __.subscriptions.dragMoveById.removeListenerForId(id, fn);
    },
    subscribeToDragEnd(id, fn) {
      __.subscriptions.dragEndById.addListenerForId(id, fn);
    },
    unsubscribeToDragEnd(id, fn) {
      __.subscriptions.dragEndById.removeListenerForId(id, fn);
    },
  };
}

function handleOnSelectStart() {
  return false;
}

interface Props {
  children?: ReactNode;
}

export function DragManagerProvider(props: Props) {
  const dragManager = useMemo(() => createDragManager(), []);
  const containerRef = useRef();
  const {children} = props;
  const style = {
    width: `100%`,
    height: `100%`,
  };

  return (
    <Context.Provider value={dragManager}>
      <div
        ref={containerRef}
        style={style}
        onMouseUp={dragManager.handleDragEnd}
        onMouseLeave={dragManager.handleDragEnd}
        onMouseDown={dragManager.handleDragStart}
      >
        {children}
      </div>
    </Context.Provider>
  );
}

export function useDragManager() {
  return useContext(Context);
}
