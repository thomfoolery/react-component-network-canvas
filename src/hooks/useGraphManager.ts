import { useContext } from "react";

import { GraphManagerContext } from "@component/contexts";
import * as Types from "@component/types";

function useGraphManager(): Types.GraphManager {
  return useContext(GraphManagerContext);
}

export { useGraphManager };
