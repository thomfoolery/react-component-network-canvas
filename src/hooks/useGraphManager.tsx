import React, {
  createContext,
  useMemo,
  useContext,
  useEffect,
  ReactNode,
} from "react";

import {createPublisher, svgGeneratePath, roundToGrid} from "@component/utils";
import {
  createOptions,
  useDragManager,
  useOptions,
  useBridge,
} from "@component/hooks";
import {DRAFT_EDGE_ID} from "@component/constants";
import * as Types from "@component/types";

import {v1 as generateUuid} from "uuid";

const Context = createContext();

function getObjectsByIdHash(objects) {
  return objects.reduce((acc, object) => {
    acc[object.id] = object;
    return acc;
  }, {});
}

function getEdgesByNodeIdHash(edges) {
  return edges.reduce((acc, edge) => {
    const {from, to} = edge;

    if (acc[from.nodeId]) {
      acc[from.nodeId].push(edge);
    } else {
      acc[from.nodeId] = [edge];
    }

    if (acc[to.nodeId]) {
      acc[to.nodeId].push(edge);
    } else {
      acc[to.nodeId] = [edge];
    }

    return acc;
  }, {});
}
interface GraphManagerPrivateProps {
  nodes: Types.Node[];
  edges: Types.Edge[];
  nodesByIdHash: {[id: string]: Types.Node};
  edgesByIdHash: {[id: string]: Types.Edge};
  edgesByNodeIdHash: {[id: string]: Types.Edge[]};
  selectedNodeIds: string[];
  bridge?: Types.Bridge;
  options: Types.Options;
  dragManager?: Types.DragManager;
  workspace?: Types.Workspace;
  subscriptions: {
    nodePositionChangeById: Types.Publisher;
    isSelectedById: Types.Publisher;
    dragDeltaById: Types.Publisher;
    nodesChange: Types.Publisher;
    edgesChange: Types.Publisher;
  };
}

interface GraphManagerArguments {
  options?: Types.Options;
  nodes?: Types.Node[];
  edges?: Types.Edge[];
  bridge?: Types.Bridge;
  dragManager?: any;
}

function createGraphManager({
  options = createOptions(),
  nodes = [],
  edges = [],
  bridge,
  dragManager,
}: GraphManagerArguments = {}): Types.GraphManager {
  const __: GraphManagerPrivateProps = {
    nodes: [],
    edges: [],
    nodesByIdHash: {},
    edgesByIdHash: {},
    edgesByNodeIdHash: {},
    selectedNodeIds: [],
    bridge,
    options,
    dragManager,
    workspace: undefined,
    subscriptions: {
      nodePositionChangeById: createPublisher(),
      isSelectedById: createPublisher(),
      dragDeltaById: createPublisher(),
      nodesChange: createPublisher(),
      edgesChange: createPublisher(),
    },
  };

  setNodes(nodes);
  setEdges(edges);

  function setNodes(nodes: Types.Node[]) {
    __.nodes = [...nodes];
    __.nodesByIdHash = getObjectsByIdHash(__.nodes);
    __.subscriptions.nodesChange.notifyAll(__.nodes);
  }

  function setEdges(edges: Types.Edge[]) {
    __.edges = [...edges];
    __.edgesByIdHash = getObjectsByIdHash(__.edges);
    __.edgesByNodeIdHash = getEdgesByNodeIdHash(__.edges);
    __.subscriptions.edgesChange.notifyAll(__.edges);
  }

  function setSelectedNodeIds(selectedNodeIds: string[]) {
    const unselectedNodeIds = __.selectedNodeIds.filter(
      (id) => !selectedNodeIds.includes(id)
    );
    const newSelectedNodeIds = selectedNodeIds.filter(
      (id) => !__.selectedNodeIds.includes(id)
    );

    __.selectedNodeIds = [...selectedNodeIds];

    requestAnimationFrame(() => {
      __.subscriptions.isSelectedById.notifyIds(newSelectedNodeIds, true);
      __.subscriptions.isSelectedById.notifyIds(unselectedNodeIds, false);
      __.bridge?.onChangeSelectedNodeIds(selectedNodeIds, API);
    });
  }

  function updateNodePositionById(id: string, dragDelta: Types.Position) {
    const {isSnapToGridEnabled} = options;
    const node = __.nodesByIdHash[id];
    const position = isSnapToGridEnabled
      ? roundToGrid(
          {
            x: node.position.x + dragDelta.x,
            y: node.position.y + dragDelta.y,
          },
          options.gridSize
        )
      : {
          x: node.position.x + dragDelta.x,
          y: node.position.y + dragDelta.y,
        };

    node.position = position;
    __.subscriptions.nodePositionChangeById.notifyIds([id], position);
  }

  function clearDraftEdgePath() {
    const svgPath = document.querySelector(`#Edge-${DRAFT_EDGE_ID}`);

    svgPath?.setAttribute("d", "");
    svgPath?.nextElementSibling?.setAttribute("d", "");
  }

  function updateDraftEdgePath(x1: number, y1: number, x2: number, y2: number) {
    const svgPath = document.querySelector(`#Edge-${DRAFT_EDGE_ID}`);
    const path = svgGeneratePath(x1, y1, x2, y2);

    svgPath?.setAttribute("d", path);
    svgPath?.nextElementSibling?.setAttribute("d", path);
  }

  const API = {
    // bridge
    get bridge() {
      return {...__.bridge};
    },
    set bridge(bridge: Types.Bridge) {
      __.bridge = bridge;
    },
    // workspace
    get workspace() {
      return {...__.workspace};
    },
    set workspace(workspace: Types.Workspace) {
      __.workspace = workspace;
    },
    // nodes
    get nodes() {
      return [...__.nodes];
    },
    set nodes(nodes: Types.Node[]) {
      setNodes(nodes);
    },
    getNodeById(id: string): Types.Node {
      return {...__.nodesByIdHash[id]};
    },
    getNodesByEdgeId(id: string): {from?: Types.Node; to?: Types.Node} {
      const edge: Types.Edge = __.edgesByIdHash[id];
      if (edge) {
        const from = __.nodesByIdHash[edge.from.nodeId];
        const to = __.nodesByIdHash[edge.to.nodeId];
        return {from, to};
      }
      return {from: undefined, to: undefined};
    },
    createNode(nodeProps: Partial<Types.Node>): Types.Node {
      const {isSnapToGridEnabled, gridSize} = options;
      const position = isSnapToGridEnabled
        ? roundToGrid(nodeProps.position || {x: 0, y: 0}, gridSize)
        : nodeProps.position || {x: 0, y: 0};

      const node = {
        id: generateUuid(),
        inputPorts: [{id: generateUuid()}],
        outputPorts: [{id: generateUuid()}],
        ...nodeProps,
        position,
      };

      setNodes([...__.nodes, node]);

      __.bridge?.onMutateGraph({
        action: "CREATE_NODE",
        subject: node,
        graph: {
          nodes: [...__.nodes],
          edges: [...__.edges],
        },
      });

      return node;
    },
    removeNodeById(id: string) {
      const edges = __.edgesByNodeIdHash[id] || [];
      const removedEdgeIds = edges.map(({id}) => id);

      setSelectedNodeIds(
        __.selectedNodeIds.filter((selectedNodeId) => selectedNodeId !== id)
      );
      setEdges(__.edges.filter((edge) => !removedEdgeIds.includes(edge.id)));
      setNodes(__.nodes.filter((node) => node.id != id));

      __.bridge?.onMutateGraph({
        action: "DELETE_NODE",
        subject: __.nodesByIdHash[id],
        graph: {
          nodes: [...__.nodes],
          edges: [...__.edges],
        },
      });
    },
    removeNodesByIds(removedNodeIds: string[]) {
      const removedEdgeIds: string[] = removedNodeIds.reduce(
        (acc: string[], id: string) => {
          const edges = __.edgesByNodeIdHash[id] || [];
          return [...acc, ...edges.map(({id}) => id)];
        },
        []
      );

      setSelectedNodeIds(
        __.selectedNodeIds.filter(
          (selectedNodeId) => !removedNodeIds.includes(selectedNodeId)
        )
      );
      setEdges(__.edges.filter((edge) => !removedEdgeIds.includes(edge.id)));
      setNodes(__.nodes.filter((node) => !removedNodeIds.includes(node.id)));

      removedNodeIds.forEach((id) => {
        __.bridge?.onMutateGraph({
          action: "DELETE_NODE",
          subject: __.nodesByIdHash[id],
          graph: {
            nodes: __.nodes,
            edges: __.edges,
          },
        });
      });
    },
    subscribeToNodesChange(fn: () => void) {
      __.subscriptions.nodesChange.addListenerForId("default", fn);
    },
    unsubscribeToNodesChange(fn: () => void) {
      __.subscriptions.nodesChange.removeListenerForId("default", fn);
    },
    // position
    set dragManager(dragManager: any) {
      __.dragManager = dragManager;
    },
    handleDragMove(event, dragDelta: Types.Position, dragData: any) {
      const {selectedNodeIds, dragManager, workspace} = __;

      if (dragManager?.dragData?.dragType === "node") {
        __.subscriptions.dragDeltaById.notifyIds(selectedNodeIds, dragDelta);
      }

      if (workspace && dragManager?.dragData?.dragType === "port") {
        const position = workspace.getCanvasPosition(event);

        const x1 = dragData.port.position.x;
        const y1 = dragData.port.position.y;
        const x2 = position.x;
        const y2 = position.y;

        updateDraftEdgePath(x1, y1, x2, y2);
      }
    },
    handleDragEnd(event, dragDelta: Types.Position, dragData: any) {
      const {selectedNodeIds, subscriptions} = __;

      subscriptions.dragDeltaById.notifyIds(selectedNodeIds, {
        x: 0,
        y: 0,
      });

      if (dragData?.dragType === "node") {
        selectedNodeIds.forEach((id) => updateNodePositionById(id, dragDelta));
      }
      if (dragData?.dragType === "port") {
        clearDraftEdgePath();
      }
    },
    updateNodePositionById,
    subscribeToDragDeltaById(id: string, fn: () => void) {
      __.subscriptions.dragDeltaById.addListenerForId(id, fn);
    },
    unsubscribeToDragDeltaById(id: string, fn: () => void) {
      __.subscriptions.dragDeltaById.removeListenerForId(id, fn);
    },
    subscribeToNodePositionChangeById(id: string, fn: () => void) {
      __.subscriptions.nodePositionChangeById.addListenerForId(id, fn);
    },
    unsubscribeToNodePositionChangeById(id: string, fn: () => void) {
      __.subscriptions.nodePositionChangeById.addListenerForId(id, fn);
    },
    // selected ids
    get selectedNodeIds(): string[] {
      return [...__.selectedNodeIds];
    },
    set selectedNodeIds(selectedNodeIds: string[]) {
      setSelectedNodeIds(selectedNodeIds);
    },
    appendSelectedNodeId(id: string) {
      if (__.selectedNodeIds.includes(id)) return;
      const selectedNodeIds = Array.from(new Set([...__.selectedNodeIds, id]));

      setSelectedNodeIds(selectedNodeIds);
    },
    appendSelectedNodeIds(appendedNodeIds: string[]) {
      const selectedNodeIds = Array.from(
        new Set([...__.selectedNodeIds, ...appendedNodeIds])
      );

      setSelectedNodeIds(selectedNodeIds);
    },
    removeSelectedNodeId(id: string) {
      if (!__.selectedNodeIds.includes(id)) return;
      const selectedNodeIds = __.selectedNodeIds.filter(
        (selectedNodeId) => selectedNodeId !== id
      );

      setSelectedNodeIds(selectedNodeIds);
    },
    removeSelectedNodeIds(unselectedNodeIds: string[]) {
      const selectedNodeIds = __.selectedNodeIds.filter(
        (id) => !unselectedNodeIds.includes(id)
      );

      setSelectedNodeIds(selectedNodeIds);
    },
    subscribeToIsSelectedById(id: string, fn: () => void) {
      __.subscriptions.isSelectedById.addListenerForId(id, fn);
    },
    unsubscribeToIsSelectedById(id: string, fn: () => void) {
      __.subscriptions.isSelectedById.removeListenerForId(id, fn);
    },
    // edges
    get edges(): Types.Edge[] {
      return [...__.edges];
    },
    set edges(edges: Types.Edge[]) {
      setEdges(edges);
    },
    getEdgeById(id: string): Types.Edge {
      return {...__.edgesByIdHash[id]};
    },
    getEdgesByNodeId(id: string): Types.Edge[] {
      return __.edgesByNodeIdHash[id] ? [...__.edgesByNodeIdHash[id]] : [];
    },
    createEdge(edgeProps: Partial<Types.Edge>): Types.Edge {
      const edge: Types.Edge = {
        id: generateUuid(),
        from: {
          nodeId: "",
          portId: "",
        },
        to: {
          nodeId: "",
          portId: "",
        },
        ...edgeProps,
      };

      setEdges([...__.edges, edge]);

      // trigger render to draw edges
      updateNodePositionById(edge.from.nodeId, {x: 0, y: 0});

      __.bridge?.onMutateGraph({
        action: "CREATE_EDGE",
        subject: edge,
        graph: {
          nodes: [...__.nodes],
          edges: [...__.edges],
        },
      });

      return edge;
    },
    removeEdgeById(id: string) {
      const filter = (edge) => edge.id != id;
      const {edgesByNodeIdHash} = __;
      const edge = __.edgesByIdHash[id];
      const fromNodeEdges = edgesByNodeIdHash[edge.from.nodeId];
      const toNodeEdges = edgesByNodeIdHash[edge.to.nodeId];

      edgesByNodeIdHash[edge.from.nodeId] = fromNodeEdges.filter(filter);
      edgesByNodeIdHash[edge.to.nodeId] = toNodeEdges.filter(filter);

      setEdges(__.edges.filter(filter));

      // redraw from and to nodes
      updateNodePositionById(edge.from.nodeId, {x: 0, y: 0});
      updateNodePositionById(edge.to.nodeId, {x: 0, y: 0});

      __.bridge?.onMutateGraph({
        action: "DELETE_EDGE",
        subject: edge,
        graph: {
          nodes: [...__.nodes],
          edges: [...__.edges],
        },
      });
    },
    clearDraftEdgePath,
    updateDraftEdgePath,
    subscribeToEdgesChange(id: string, fn: () => void) {
      __.subscriptions.edgesChange.addListenerForId(id, fn);
    },
    unsubscribeToEdgesChange(id: string, fn: () => void) {
      __.subscriptions.edgesChange.removeListenerForId(id, fn);
    },
    // import/export
    import({nodes, edges}) {
      setNodes(nodes);
      setEdges(edges);
    },
    export() {
      return {
        nodes: [...__.nodes],
        edges: [...__.edges],
      };
    },
  };

  return API;
}

interface Props {
  nodes: Types.Node[];
  edges: Types.Edge[];
  children?: ReactNode;
}

function GraphManagerProvider(props: Props) {
  const {nodes, edges, children} = props;
  const dragManager = useDragManager();
  const options = useOptions();
  const bridge = useBridge();

  const graphManager = useMemo(
    () => createGraphManager({options, nodes, edges, bridge, dragManager}),
    []
  );

  useEffect(() => {
    const id = "graphManager";

    bridge.connect(graphManager);
    dragManager.subscribeToDragMove(id, graphManager.handleDragMove);
    dragManager.subscribeToDragEnd(id, graphManager.handleDragEnd);

    return () => {
      dragManager.unsubscribeToDragMove(id, graphManager.handleDragMove);
      dragManager.unsubscribeToDragEnd(id, graphManager.handleDragEnd);
    };
  }, [bridge, dragManager]);

  return <Context.Provider value={graphManager}>{children}</Context.Provider>;
}

function useGraphManager() {
  return useContext(Context);
}

export {createGraphManager, useGraphManager, GraphManagerProvider};
