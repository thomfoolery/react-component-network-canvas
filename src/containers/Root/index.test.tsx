import React from "react";
import renderer from "react-test-renderer";
import { createOptions } from "@component/hooks";

import { Root } from "./index";

it("renders correctly", () => {
  const nodes = [
    {
      id: "node-1",
      position: {
        x: 0,
        y: 0,
      },
      inputPorts: [{ id: "node-1-input-1" }],
      outputPorts: [{ id: "node-1-output-1" }],
    },
    {
      id: "node-2",
      position: {
        x: 100,
        y: 100,
      },
      inputPorts: [{ id: "node-2-input-1" }],
      outputPorts: [{ id: "node-2-output-1" }],
    },
  ];
  const edges = [
    {
      id: "edge-1",
      from: {
        nodeId: "node-1",
        portId: "node-1-output-1",
      },
      to: {
        nodeId: "node-2",
        portId: "node-2-input-1",
      },
    },
  ];
  const options = createOptions();
  const tree = renderer
    .create(<Root nodes={nodes} edges={edges} options={options} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
