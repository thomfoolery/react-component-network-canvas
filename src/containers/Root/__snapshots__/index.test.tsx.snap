// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "--NC-canvas-background-color": "white",
        "--NC-canvas-background-image": "radial-gradient(lightgray, lightgray 1px, transparent 1px)",
        "--NC-canvas-background-position": "50% 50%",
        "--NC-canvas-background-size": "var(--NC-grid-size) var(--NC-grid-size)",
        "--NC-canvas-border-radius": "5px",
        "--NC-canvas-box-shadow": "0 0 0 1px lightgrey",
        "--NC-canvas-size": "undefinedpx",
        "--NC-edge-draft-stroke": "black",
        "--NC-edge-hover-stroke": "red",
        "--NC-edge-stroke": "black",
        "--NC-edge-stroke-width": "3px",
        "--NC-grid-size": "undefinedpx",
        "--NC-workspace-background-color": "#f6f6f6",
        "--NC-workspace-background-image": "",
        "--NC-workspace-background-size": "",
      }
    }
  >
    <div
      onMouseDown={[Function]}
    >
      <div
        onClick={[Function]}
        onMouseDown={[Function]}
        style={
          Object {
            "transform": "translate3d(0px,0px,0) scale(1)",
          }
        }
      >
        <div />
        <svg
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <path
              id="Edge-edge-1"
            />
            <path />
          </g>
          <g
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <path
              id="Edge-e3327200-8456-11eb-ad94-117561157804"
            />
            <path />
          </g>
        </svg>
        <div
          className=""
          id="Node-node-1"
          style={
            Object {
              "left": 0,
              "top": 0,
            }
          }
        >
          <div>
            <div
              id="Port-node-1-input-1"
              onMouseDown={[Function]}
              onMouseUp={[Function]}
            />
          </div>
          <div
            onMouseDown={[Function]}
            onMouseUp={[Function]}
          />
          <div>
            <div
              id="Port-node-1-output-1"
              onMouseDown={[Function]}
              onMouseUp={[Function]}
            />
          </div>
        </div>
        <div
          className=""
          id="Node-node-2"
          style={
            Object {
              "left": 100,
              "top": 100,
            }
          }
        >
          <div>
            <div
              id="Port-node-2-input-1"
              onMouseDown={[Function]}
              onMouseUp={[Function]}
            />
          </div>
          <div
            onMouseDown={[Function]}
            onMouseUp={[Function]}
          />
          <div>
            <div
              id="Port-node-2-output-1"
              onMouseDown={[Function]}
              onMouseUp={[Function]}
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
